src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/SwerveStateHelper.java

```diff
@@ -0,0 +1,119 @@
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.math.kinematics.ChassisSpeeds;
+import frc.robot.constants.MathConstants;
+import frc.robot.subsystems.swerve.Swerve;
+import frc.robot.subsystems.swerve.SwerveConstants;
+import frc.robot.subsystems.swerve.SwerveMath;
+import frc.robot.subsystems.swerve.modules.ModuleUtils;
+
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static frc.robot.subsystems.swerve.swervestatehelpers.AimAssistUtils.getRotationAssistedSpeeds;
+
+public class SwerveStateHelper {
+
+    private final Supplier<Pose2d> robotPoseSupplier;
+    private final Swerve swerve;
+    private final SwerveConstants swerveConstants;
+    private final Supplier<Optional<Translation2d>> noteTranslationSupplier;
+
+    public SwerveStateHelper(Supplier<Pose2d> robotPoseSupplier, Supplier<Optional<Translation2d>> noteTranslationSupplier,
+            Swerve swerve) {
+        this.robotPoseSupplier = robotPoseSupplier;
+        this.swerve = swerve;
+        this.swerveConstants = swerve.getConstants();
+        this.noteTranslationSupplier = noteTranslationSupplier;
+    }
+
+    public ChassisSpeeds applyAimAssistOnSpeeds(AimAssist aimAssistState, ChassisSpeeds speeds) {
+        return switch (aimAssistState) {
+            case SPEAKER -> getRotationAssistedSpeeds(
+                    speeds,
+                    swerve.getRobotRelativeVelocity(),
+                    () -> robotPoseSupplier.get().getRotation(),
+                    () -> SwerveMath.getRelativeTranslation(robotPoseSupplier.get().getTranslation(),new Translation2d(0,0)).getAngle(),
+                    swerveConstants
+            );
+            case AMP -> getRotationAssistedSpeeds(
+                    speeds,
+                    swerve.getRobotRelativeVelocity(),
+                    () -> robotPoseSupplier.get().getRotation(),
+                    () -> Rotation2d.fromDegrees(90),
+                    swerveConstants
+            );
+            case NOTE -> handleNoteAssistState(
+                    speeds,
+                    robotPoseSupplier,
+                    noteTranslationSupplier
+            );
+            case NONE -> speeds;
+        };
+    }
+
+
+    private ChassisSpeeds handleNoteAssistState(
+            ChassisSpeeds speeds,
+            Supplier<Pose2d> robotPoseSupplier,
+            Supplier<Optional<Translation2d>> noteTranslationSupplier) {
+        if (noteTranslationSupplier.get().isEmpty()) {
+            return speeds;
+        }
+
+        Translation2d noteRelativeToRobot = SwerveMath.getPoseRelativeTranslation(robotPoseSupplier.get(),
+                noteTranslationSupplier.get().get());
+
+
+        double pidGainHorizontalVelocity = swerveConstants.yMetersPIDController().calculate(
+                0,
+                noteRelativeToRobot.getY()
+        );
+
+        double wantedHorizontalVelocity = pidGainHorizontalVelocity + speeds.vyMetersPerSecond;
```

Comment:
the speed.y is field relative not robot like it can be both

Reply:
its horizontal in robot oriented

Reply:
i mean the speeds.vyMetersPerSecond and pidGainHorizontalVelocity  are not same horizontal